### PR TITLE
HPCC-25854 WsSQL nested dataset support should be versioned

### PR DESCRIPTION
**Pull request by rpastrana**
- Versions nested dataset related interface changes
- Improves logic rejecting normalized column arithmatic

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by rpastrana** [comment]
@wangkx please review. https://github.com/hpcc-systems/HPCC-Platform/pull/14644 should have been version aware

**Comment by rpastrana** [comment]
@Michael-Gardner once this PR is merged, we'll need to target this HPCC version in the hpcc4j repo to avoid false negative junit failures

File: esp/services/ws_sql/SQL2ECL/HPCCFileCache.cpp [hunk 1]
```diff
@@ -80,12 +80,15 @@ bool HPCCFileCache::populateTablesResponse(IEspGetDBMetaDataResponse & tablesres
...
                    HPCCColumnMetaData& currcol = cols->item(i);
 
-                   IArrayOf<HPCCColumnMetaData> * ccols = currcol.getChildColumns();
-                   if (ccols->length() != 0)
+                   if (nestedcolumns)
                    {
-                       IArrayOf<IEspHPCCColumn> pCColumns;
-                       populateColums(ccols, pCColumns);
-                       pCol->setColumns(pCColumns);
+					   IArrayOf<HPCCColumnMetaData> * ccols = currcol.getChildColumns();
```
Comment (wangkx) [inline esp/services/ws_sql/SQL2ECL/HPCCFileCache.cpp:85]: wrong indent

File: esp/services/ws_sql/SQL2ECL/HPCCSQLTreeWalker.cpp [hunk 2]
```diff
@@ -304,12 +304,20 @@ ISQLExpression * HPCCSQLTreeWalker::expressionTreeWalker(pANTLR3_BASE_TREE exprA
             case LIKE_SYM:
             case NOT_LIKE:
                 leftexp.set(expressionTreeWalker((pANTLR3_BASE_TREE)(exprAST->getChild(exprAST, 0)),exprAST));
-                if (leftexp->getExpType() != Value_ExpressionType && strnicmp(leftexp->getECLType(),"DATASET",7)==0)
-                    throw MakeStringException(-1, "Cannot apply arithmetic logic to normalized nested column: '%s'!\n", leftexp->getName());
+                if (leftexp->getLogicType() == NestedDS_LogicType)
+                {
+                    StringBuffer invalidexp;
+                    leftexp->toString(invalidexp, false);
+                    throw MakeStringException(-1, "Cannot apply arithmetic logic to normalized nested column: '%s'!\n", invalidexp.str());
```
Comment (wangkx) [inline esp/services/ws_sql/SQL2ECL/HPCCSQLTreeWalker.cpp:311]: Please use makeStringExceptionV

File: esp/services/ws_sql/SQL2ECL/HPCCSQLTreeWalker.cpp [hunk 3]
```diff
@@ -304,12 +304,20 @@ ISQLExpression * HPCCSQLTreeWalker::expressionTreeWalker(pANTLR3_BASE_TREE exprA
...
+                }
 
                 rightexp.set(expressionTreeWalker((pANTLR3_BASE_TREE)(exprAST->getChild(exprAST, 1)),exprAST));
-                if (rightexp->getExpType() != Value_ExpressionType && strnicmp(rightexp->getECLType(),"DATASET",7)==0)
-                    throw MakeStringException(-1, "Cannot apply arithmetic logic to normalized nested column: '%s'!\n", rightexp->getName());
+                if (rightexp->getLogicType() == NestedDS_LogicType)
+                {
+                    StringBuffer invalidexp;
+                    rightexp->toString(invalidexp, false);
+                    throw MakeStringException(-1, "Cannot apply arithmetic logic to normalized nested column: '%s'!\n", invalidexp.str());
```
Comment (wangkx) [inline esp/services/ws_sql/SQL2ECL/HPCCSQLTreeWalker.cpp:319]: Please use makeStringExceptionV

File: esp/services/ws_sql/SQL2ECL/HPCCSQLTreeWalker.cpp [hunk 4]
```diff
@@ -341,8 +349,12 @@ ISQLExpression * HPCCSQLTreeWalker::expressionTreeWalker(pANTLR3_BASE_TREE exprA
             case NOT_SYM:
             {
                 tmpexp.setown(new SQLUnaryExpression(expressionTreeWalker((pANTLR3_BASE_TREE)(exprAST->getChild(exprAST, 0)),exprAST), exptype ));
-                if (tmpexp->getExpType() != Value_ExpressionType && strnicmp(tmpexp->getECLType(),"DATASET",7)==0)
-                    throw MakeStringException(-1, "Cannot apply arithmetic logic to normalized nested column: '%s'!\n", rightexp->getName());
+                if (tmpexp->getLogicType() == NestedDS_LogicType)
+                {
+                    StringBuffer expstr;
+                    tmpexp->toString(expstr, false);
+                    throw MakeStringException(-1, "Cannot apply arithmetic logic to normalized nested column: '%s'!\n", expstr.str());
```
Comment (wangkx) [inline esp/services/ws_sql/SQL2ECL/HPCCSQLTreeWalker.cpp:356]: Please use makeStringExceptionV

File: esp/services/ws_sql/ws_sqlService.cpp [hunk 5]
```diff
@@ -53,7 +53,7 @@ bool CwssqlEx::onGetDBMetaData(IEspContext &context, IEspGetDBMetaDataRequest &r
     if (includetables)
     {
         Owned<HPCCFileCache> tmpHPCCFileCache = HPCCFileCache::createFileCache(username.str(), passwd);
-        tmpHPCCFileCache->populateTablesResponse(resp, req.getTableFilter());
+        tmpHPCCFileCache->populateTablesResponse(resp, req.getTableFilter(), context.getClientVersion() > 3.05);
```
Comment (wangkx) [inline esp/services/ws_sql/ws_sqlService.cpp:56]: Better use '>= 3.06'

**Review by wangkx — CHANGES_REQUESTED** [review]
@rpastrana please see my minor comments

**Review by wangkx — APPROVED** [review]
@rpastrana looks fine.

**Pull request merged**